### PR TITLE
[ExternalInterfaces] Make fill non-hoistableLeafOp, hoist linalg init operands

### PR DESCRIPTION
**Pull request by Max191**
This PR changes the HoistableOpInterface implementation for LinalgOps to make DPS init operands hoistable. If there is real computation on the init of a linalg op, then we would want to hoist that computation. The main way the current policy (not hoisting init operands) affects hoisting today is by preventing linalg.fill ops from being hoisted into splat constants, since the init operands are usually linalg.fill or tensor.empty. Since we want to preserve the behavior of not hoisting fills (and instead fusing them with consumers to handle in codegen), the interface implementation now treats fill ops as non-HoistableLeafOps.

This PR disables the top-k test on rocm, since it exposed a bug in rocm codegen, but we want to move progress forward on GPU data tiling. The test should be added back once the bug is resolved.

### CODE REVIEW

File: compiler/src/iree/compiler/ExternalInterfaces/UtilExternalModels.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/ExternalInterfaces/UtilExternalModels.cpp:232]: nit: I think there is a `linalg::isaFillOpInterface`

**Review by qedawkins — APPROVED** [review]
Looks good, I assume this is already tested by existing tests?

**Comment by Max191** [comment]
This breaks the top_k e2e test, but it looks like the top_k lowering was broken to begin with, and constant folding the fills was "fixing" it by adding extra barriers at bufferization. I'm going to need to fix this before landing.

File: compiler/src/iree/compiler/ExternalInterfaces/UtilExternalModels.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/ExternalInterfaces/UtilExternalModels.cpp:227]: style nit: adding a blank line before the comment usually helps the readability.

see https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace for more details.

**Review by hanhanW — CHANGES_REQUESTED** [review]
Thanks for taking care of this! I'd like to take an additional review after the topk issue is fixed, so I put a "request changes" for now.

**Comment by Max191** [comment]
> Thanks for taking care of this! I'd like to take an additional review after the topk issue is fixed, so I put a "request changes" for now.

I talked to Mahesh offline and we decided it would be okay to open this in an issue and disable the test for now in this PR. @krzysz00 offered to take a look when he has some time, but it might take a few days. Since topK is not very important we decided it is okay to disable the test to unblock GPU data tiling progress. WDYT?

EDIT: issue is here: https://github.com/iree-org/iree/issues/18649

**Review by hanhanW — APPROVED** [review]
SGTM

**Pull request merged**
